Changing the implementation of DeepCopy to use reflection

krousey opened the pull request:
The Gob encode/decode method did not properly handle pointers to primitive types as discussed in #8054.

Commit:
Changing the implementation of DeepCopy to use reflection instead of Gob encode/decode.

krousey commented:
There was a timeout waiting for a POD in the shippable build. I requested to run the build again. It passed unit, integration, and e2e when I ran it.

krousey commented:
@smarterclayton Would you mind taking a look at this one?

krousey commented:
@lavalamp @wojtek-t FYI.

smarterclayton commented:
Is there a way to put the deep copier into api/serialization_test.go's code path so it's run on everything?  Probably in roundTrip()

krousey commented:
Would it make more sense to put this in deep_copy's test and not the api's test?

krousey commented:
I guess that does create a import loop. I'd still rather just write a separate test function for it.

smarterclayton commented:
Sure - a fuzzer that goes through all the api objects is fine.  You'll want multiple iterations.  But that is SGTM

----- Original Message -----
> I guess that does create a import loop. I'd still rather just write a
> separate test function for it.
> 
> ---
> Reply to this email directly or view it on GitHub:
> https://github.com/GoogleCloudPlatform/kubernetes/pull/8188#issuecomment-101836144

krousey commented:
Added an API object test. I then tried to break it to see that it wasn't breaking. Apparently Value.CanSet() returns false for embedded members when created with reflect.Zero(). So I changed it to use reflect.New() to get a pointer and access it through Elem().

smarterclayton commented on pkg/api/copy_test.go:
```diff
@@ -0,0 +1,47 @@
+func TestDeepCopyApiObjects(t *testing.T) {
+	f := apitesting.FuzzerFor(t, "", rand.NewSource(rand.Int63()))
+	for kind := range api.Scheme.KnownTypes("") {
+		item, err := api.Scheme.New("", "Pod")
```
Should be kind?

Also, I would recommend checking the external versions as well, and running the whole test 5-20 times

krousey replied:
Thanks for catching that. I also added a loop, but I'm unsure what you mean by external versions as well. I'm a little new to this project.

smarterclayton replied:
Just add a loop with:

```
for _, version := range []string{"", testapi.Version()} {
  for kind := range api.Scheme.KnownTypes(version) {
    item, err := api.Scheme.New(version, kind)
    ...
  }
}
```

That will test the internal types always, and then whatever external type is defaulted (hack/test-go.sh tests several API versions in a row, so this test will be called multiple times with different testapi.Version() values).

----- Original Message -----
> > +package api_test
> > +
> > +import (
> > +	"math/rand"
> > +	"reflect"
> > +	"testing"
> > +
> > +	"github.com/GoogleCloudPlatform/kubernetes/pkg/api"
> > +	apitesting "github.com/GoogleCloudPlatform/kubernetes/pkg/api/testing"
> > +	"github.com/GoogleCloudPlatform/kubernetes/pkg/conversion"
> > +)
> > +
> > +func TestDeepCopyApiObjects(t *testing.T) {
> > +	f := apitesting.FuzzerFor(t, "", rand.NewSource(rand.Int63()))
> > +	for kind := range api.Scheme.KnownTypes("") {
> > +		item, err := api.Scheme.New("", "Pod")
> 
> Thanks for catching that. I also added a loop, but I'm unsure what you mean
> by external versions as well. I'm a little new to this project.
> 
> ---
> Reply to this email directly or view it on GitHub:
> https://github.com/GoogleCloudPlatform/kubernetes/pull/8188/files#r30337924

krousey replied:
Done.

wojtek-t commented on pkg/conversion/deep_copy.go:
```diff
@@ -17,52 +17,104 @@ limitations under the License.
+			if err != nil {
+				return src, err
+			}
+			dst.SetMapIndex(k, copyVal)
```
This doesn't "deep-copy" keys of the map. It works for all our current api types, but this might not be true in the future, right? Maybe a TODO to fix it?

krousey replied:
Keys to a map in most cases can't be mutated, and in all cases should not. I don't think this is a big issue.

In fact, if you had a struct as a key that had a pointer, it would be hashed on what the pointer is, not the value of what it points to. So a deep copied key would no longer be equivalent.

http://play.golang.org/p/iXtIlEFz7J

wojtek-t commented:
Thanks for this change - generally looks good to me.

smarterclayton commented:
LGTM.  Will merge on green.

krousey commented:
Should I squash the commits now?

smarterclayton commented:
Yes, I don't know why shippable hasn't completed yet though.

smarterclayton commented:
Got one shippable green, would like one more.

krousey commented:
Should I get out and push?

smarterclayton commented:
If you have a bat that might help.

vmarmol commented:
Green, merging!
